### PR TITLE
Support esm externals in app router

### LINKED ISSUE #60756: Server actions: Cannot use classes imported from ESM packages included in `serverComponentsExternalPackages` option
### Link to the code that reproduces this issue

https://github.com/pilcrowOnPaper/next-server-actions-esm-class-repro

### To Reproduce

1. Start the dev server
2. (Click the button on page `/`)
3. Observe the error in the console

### Current vs. Expected behavior

I can't seem to use classes imported from ESM packages included in experimental `serverComponentsExternalPackages` option in server actions, when the action is imported from client components. 

This works fine when the package is CJS, when the package is not included in `serverComponentsExternalPackages`, or when the action is passed down to the client component as a prop.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: x64
  Version: Darwin Kernel Version 23.2.0: Wed Nov 15 21:54:10 PST 2023; root:xnu-10002.61.3~2/RELEASE_X86_64
Binaries:
  Node: 20.2.0
  npm: 9.6.6
  Yarn: 1.22.19
  pnpm: 8.10.2
Relevant Packages:
  next: 14.0.5-canary.60
  eslint-config-next: 14.0.4
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router, Module resolution (CJS / ESM, module resolving)

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local)

### Additional context

I need to use `serverComponentsExternalPackages` as certain modules (`oslo/password`) require Node specific modules (`@node-rs/argon2`)

<sub>[NEXT-2435](https://linear.app/vercel/issue/NEXT-2435/server-actions-cannot-use-classes-imported-from-esm-packages-included)</sub>

### PR DESCRIPTION
**Pull request by huozhi**
### What

Support `esmExternals` working in app router

### Why

`esmExternals` was disabled for app router that most of the packages are picking up the CJS bundles for externals. This PR enables to resolve the ESM bundle for external packages.

We have two issues discovered while enabling the flag, any esm external packages will fail in client page SSR and server action. We fixed them by changing the below bundling logics:

* When a client page having a async dependency, we can await the page during in rendering
* When a server action having a async dependency, we changed the server action entry creation with webpack along with the server client entry creation together, then webpack can handle the modules async propagation properly.

Fixes #60756 
Closes NEXT-2435
Closes NEXT-2472
Closes NEXT-3225

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 61206c5fe016454c8f49db791d39e63f4d69d549

`__NEXT_EXPERIMENTAL_PPR=true pnpm test-start test/e2e/app-dir/actions-allowed-origins/app-action-allowed-origins.test.ts` (PPR)
- app-dir action allowed origins > should pass if localhost is set as a safe origin

<details>
<summary>Expand output</summary>

  ● app-dir action allowed origins › should pass if localhost is set as a safe origin

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

`__NEXT_EXPERIMENTAL_PPR=true pnpm test-start test/e2e/app-dir/app-external/app-external.test.ts` (PPR)
- app dir - external dependency > should be able to opt-out 3rd party packages being bundled in server components
- app dir - external dependency > should handle external async module libraries correctly
- app dir - external dependency > should transpile specific external packages with the `transpilePackages` option
- app dir - external dependency > should resolve the subset react in server components based on the react-server condition
- app dir - external dependency > should resolve 3rd party package exports based on the react-server condition
- app dir - external dependency > should correctly collect global css imports and mark them as side effects
- app dir - external dependency > should handle external css modules
- app dir - external dependency > should use the same export type for packages in both ssr and client
- app dir - external dependency > should handle external css modules in pages
- app dir - external dependency > should handle external next/font
- app dir - external dependency > should not apply swc optimizer transform for external packages in browser layer in web worker
- app dir - external dependency > should emit cjs helpers for external cjs modules when compiled
- app dir - external dependency > should export client module references in esm
- app dir - external dependency > should support exporting multiple star re-exports
- app dir - external dependency > should have proper tree-shaking for known modules in CJS
- app dir - external dependency > should use the same async storages if imported directly
- app dir - external dependency > app route > should resolve next/server api from external esm package
- app dir - external dependency > mixed syntax external modules > should handle mixed module with next/dynamic
- app dir - external dependency > mixed syntax external modules > should handle mixed module in server and client components
- app dir - external dependency > react in external esm packages > should use the same react in client app
- app dir - external dependency > react in external esm packages > should use the same react in server app
- app dir - external dependency > react in external esm packages > should use the same react in edge server app
- app dir - external dependency > react in external esm packages > should use the same react in pages
- app dir - external dependency > react in external esm packages > should support namespace import with ESM packages
- app dir - external dependency > server actions > should not prefer to resolve esm over cjs for bundling optout packages
- app dir - external dependency > server actions > should compile server actions from node_modules in client components

<details>
<summary>Expand output</summary>

  ● app dir - external dependency › should be able to opt-out 3rd party packages being bundled in server components

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › should handle external async module libraries correctly

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › should transpile specific external packages with the `transpilePackages` option

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › should resolve the subset react in server components based on the react-server condition

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › should resolve 3rd party package exports based on the react-server condition

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › should correctly collect global css imports and mark them as side effects

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › should handle external css modules

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › should use the same export type for packages in both ssr and client

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › should handle external css modules in pages

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › should handle external next/font

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › should not apply swc optimizer transform for external packages in browser layer in web worker

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › react in external esm packages › should use the same react in client app

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › react in external esm packages › should use the same react in server app

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › react in external esm packages › should use the same react in edge server app

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › react in external esm packages › should use the same react in pages

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › react in external esm packages › should support namespace import with ESM packages

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › mixed syntax external modules › should handle mixed module with next/dynamic

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › mixed syntax external modules › should handle mixed module in server and client components

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › should emit cjs helpers for external cjs modules when compiled

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › should export client module references in esm

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › should support exporting multiple star re-exports

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › should have proper tree-shaking for known modules in CJS

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › should use the same async storages if imported directly

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › server actions › should not prefer to resolve esm over cjs for bundling optout packages

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › server actions › should compile server actions from node_modules in client components

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - external dependency › app route › should resolve next/server api from external esm package

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-externals-app-router | Change |
| - | - | - | - |
| buildDuration | 16.6s | 15.1s | N/A |
| buildDurationCached | 9s | 6.9s | N/A |
| nodeModulesSize | 239 MB | 239 MB | ⚠️ +3.55 kB |
| nextStartRea..uration (ms) | 410ms | 408ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-externals-app-router | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 31.8 kB | 31.8 kB | N/A |
| 3304.HASH.js gzip | 169 B | 169 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.5 kB | 53.5 kB | N/A |
| 8299-HASH.js gzip | 5.09 kB | 5.09 kB | ✓ |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 228 B | 227 B | N/A |
| main-HASH.js gzip | 31.5 kB | 31.6 kB | N/A |
| webpack-HASH.js gzip | 1.64 kB | 1.65 kB | N/A |
| Overall change | 50.5 kB | 50.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-externals-app-router | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-externals-app-router | Change |
| - | - | - | - |
| _app-HASH.js gzip | 193 B | 194 B | N/A |
| _error-HASH.js gzip | 193 B | 191 B | N/A |
| amp-HASH.js gzip | 510 B | 510 B | ✓ |
| css-HASH.js gzip | 342 B | 343 B | N/A |
| dynamic-HASH.js gzip | 2.51 kB | 2.52 kB | N/A |
| edge-ssr-HASH.js gzip | 265 B | 265 B | ✓ |
| head-HASH.js gzip | 365 B | 364 B | N/A |
| hooks-HASH.js gzip | 389 B | 391 B | N/A |
| image-HASH.js gzip | 4.28 kB | 4.28 kB | N/A |
| index-HASH.js gzip | 269 B | 268 B | N/A |
| link-HASH.js gzip | 2.68 kB | 2.69 kB | N/A |
| routerDirect..HASH.js gzip | 328 B | 326 B | N/A |
| script-HASH.js gzip | 395 B | 397 B | N/A |
| withRouter-HASH.js gzip | 323 B | 323 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.2 kB | 1.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-externals-app-router | Change |
| - | - | - | - |
| _buildManifest.js gzip | 482 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-externals-app-router | Change |
| - | - | - | - |
| index.html gzip | 528 B | 530 B | N/A |
| link.html gzip | 541 B | 541 B | ✓ |
| withRouter.html gzip | 523 B | 523 B | ✓ |
| Overall change | 1.06 kB | 1.06 kB | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-externals-app-router | Change |
| - | - | - | - |
| edge-ssr.js gzip | 108 kB | 108 kB | N/A |
| page.js gzip | 3.05 kB | 3.04 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-externals-app-router | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 622 B | 623 B | N/A |
| middleware-r..fest.js gzip | 155 B | 156 B | N/A |
| middleware.js gzip | 27.8 kB | 27.8 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 839 B | 839 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-externals-app-router | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | N/A |
| app-page-exp..prod.js gzip | 98.3 kB | 98.3 kB | ✓ |
| app-page-tur..prod.js gzip | 99.8 kB | 99.8 kB | ✓ |
| app-page-tur..prod.js gzip | 94.2 kB | 94.2 kB | ✓ |
| app-page.run...dev.js gzip | 157 kB | 157 kB | N/A |
| app-page.run..prod.js gzip | 92.9 kB | 92.9 kB | ✓ |
| app-route-ex...dev.js gzip | 21.4 kB | 21.4 kB | ✓ |
| app-route-ex..prod.js gzip | 15.1 kB | 15.1 kB | ✓ |
| app-route-tu..prod.js gzip | 15.1 kB | 15.1 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route.ru...dev.js gzip | 21.2 kB | 21.2 kB | ✓ |
| app-route.ru..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 21.4 kB | 21.4 kB | ✓ |
| pages.runtim...dev.js gzip | 22.1 kB | 22.1 kB | ✓ |
| pages.runtim..prod.js gzip | 21.4 kB | 21.4 kB | ✓ |
| server.runti..prod.js gzip | 65.3 kB | 65.3 kB | ✓ |
| Overall change | 647 kB | 647 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix/esm-externals-app-router | Change |
| - | - | - | - |
| 0.pack gzip | 1.61 MB | 1.63 MB | ⚠️ +14.5 kB |
| index.pack gzip | 107 kB | 113 kB | ⚠️ +5.89 kB |
| Overall change | 1.72 MB | 1.74 MB | ⚠️ +20.4 kB |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>image-HASH.js</strong></summary>

```diff
@@ -1,7 +1,7 @@
 (self["webpackChunk_N_E"] = self["webpackChunk_N_E"] || []).push([
   [8358],
   {
-    /***/ 1552: /***/ (
+    /***/ 4070: /***/ (
       __unused_webpack_module,
       __unused_webpack_exports,
       __webpack_require__
@@ -9,7 +9,7 @@
       (window.__NEXT_P = window.__NEXT_P || []).push([
         "/image",
         function () {
-          return __webpack_require__(5237);
+          return __webpack_require__(396);
         },
       ]);
       if (false) {
@@ -18,7 +18,7 @@
       /***/
     },
 
-    /***/ 2016: /***/ (module, exports, __webpack_require__) => {
+    /***/ 8490: /***/ (module, exports, __webpack_require__) => {
       "use strict";
       /* __next_internal_client_entry_do_not_use__  cjs */
       Object.defineProperty(exports, "__esModule", {
@@ -40,15 +40,15 @@
         __webpack_require__(422)
       );
       const _head = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(6074)
+        __webpack_require__(2457)
       );
-      const _getimgprops = __webpack_require__(9571);
-      const _imageconfig = __webpack_require__(6567);
-      const _imageconfigcontextsharedruntime = __webpack_require__(419);
-      const _warnonce = __webpack_require__(4486);
-      const _routercontextsharedruntime = __webpack_require__(162);
+      const _getimgprops = __webpack_require__(7932);
+      const _imageconfig = __webpack_require__(5706);
+      const _imageconfigcontextsharedruntime = __webpack_require__(9483);
+      const _warnonce = __webpack_require__(9035);
+      const _routercontextsharedruntime = __webpack_require__(4829);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(6996)
+        __webpack_require__(7240)
       );
       // This is replaced by webpack define plugin
       const configEnv = {
@@ -379,7 +379,7 @@
       /***/
     },
 
-    /***/ 9571: /***/ (
+    /***/ 7932: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -395,9 +395,9 @@
           return getImgProps;
         },
       });
-      const _warnonce = __webpack_require__(4486);
-      const _imageblursvg = __webpack_require__(133);
-      const _imageconfig = __webpack_require__(6567);
+      const _warnonce = __webpack_require__(9035);
+      const _imageblursvg = __webpack_require__(2642);
+      const _imageconfig = __webpack_require__(5706);
       const VALID_LOADING_VALUES =
         /* unused pure expression or super */ null && [
           "lazy",
@@ -772,7 +772,7 @@
       /***/
     },
 
-    /***/ 133: /***/ (__unused_webpack_module, exports) => {
+    /***/ 2642: /***/ (__unused_webpack_module, exports) => {
       "use strict";
       /**
        * A shared function, used on both client and server, to generate a SVG blur placeholder.
@@ -827,7 +827,7 @@
       /***/
     },
 
-    /***/ 4085: /***/ (
+    /***/ 503: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -854,10 +854,10 @@
         },
       });
       const _interop_require_default = __webpack_require__(7456);
-      const _getimgprops = __webpack_require__(9571);
-      const _imagecomponent = __webpack_require__(2016);
+      const _getimgprops = __webpack_require__(7932);
+      const _imagecomponent = __webpack_require__(8490);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(6996)
+        __webpack_require__(7240)
       );
       function getImageProps(imgProps) {
         const { props } = (0, _getimgprops.getImgProps)(imgProps, {
@@ -889,7 +889,7 @@
       /***/
     },
 
-    /***/ 6996: /***/ (__unused_webpack_module, exports) => {
+    /***/ 7240: /***/ (__unused_webpack_module, exports) => {
       "use strict";
 
       Object.defineProperty(exports, "__esModule", {
@@ -924,7 +924,7 @@
       /***/
     },
 
-    /***/ 5237: /***/ (
+    /***/ 396: /***/ (
       __unused_webpack_module,
       __webpack_exports__,
       __webpack_require__
@@ -941,8 +941,8 @@
 
       // EXTERNAL MODULE: ./node_modules/.pnpm/react@18.2.0/node_modules/react/jsx-runtime.js
       var jsx_runtime = __webpack_require__(1527);
-      // EXTERNAL MODULE: ./node_modules/.pnpm/file+..+main-repo+packages+next+next-packed.tgz_react-dom@18.2.0_react@18.2.0/node_modules/next/image.js
-      var next_image = __webpack_require__(1577);
+      // EXTERNAL MODULE: ./node_modules/.pnpm/file+..+diff-repo+packages+next+next-packed.tgz_react-dom@18.2.0_react@18.2.0/node_modules/next/image.js
+      var next_image = __webpack_require__(73);
       var image_default = /*#__PURE__*/ __webpack_require__.n(next_image); // CONCATENATED MODULE: ./pages/nextjs.png
       /* harmony default export */ const nextjs = {
         src: "/_next/static/media/nextjs.cae0b805.png",
@@ -972,12 +972,8 @@
       /***/
     },
 
-    /***/ 1577: /***/ (
-      module,
-      __unused_webpack_exports,
-      __webpack_require__
-    ) => {
-      module.exports = __webpack_require__(4085);
+    /***/ 73: /***/ (module, __unused_webpack_exports, __webpack_require__) => {
+      module.exports = __webpack_require__(503);
 
       /***/
     },
@@ -987,7 +983,7 @@
     /******/ var __webpack_exec__ = (moduleId) =>
       __webpack_require__((__webpack_require__.s = moduleId));
     /******/ __webpack_require__.O(0, [2888, 9774, 179], () =>
-      __webpack_exec__(1552)
+      __webpack_exec__(4070)
     );
     /******/ var __webpack_exports__ = __webpack_require__.O();
     /******/ _N_E = __webpack_exports__;
```
</details>
<details>
<summary>Diff for <strong>2453-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..ntime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.dev.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 61206c5fe016454c8f49db791d39e63f4d69d549

File: test/e2e/app-dir/app-external/app/optout/action/actions.js [hunk 1]
```diff
@@ -1,7 +1,9 @@
 'use server'
 
 import { value as dualPkgOptoutValue } from 'dual-pkg-optout'
+import * as mod from 'dual-pkg-optout'
 
 export async function getDualOptoutValue() {
+  console.log('getDualOptoutValue', dualPkgOptoutValue, 'mod', mod)
```
Comment (ForsakenHarmony) [inline test/e2e/app-dir/app-external/app/optout/action/actions.js:7]: Probably need to remove this again?

File: test/e2e/app-dir/app-external/app/optout/action/page.js [hunk 2]
```diff
@@ -10,7 +10,9 @@ export default function Page() {
       <p>{optoutDisplayValue}</p>
       <button
         onClick={async () => {
-          setOptoutDisplayValue(await getDualOptoutValue())
+          const value = await getDualOptoutValue()
+          console.log('value', value, 'from', getDualOptoutValue)
```
Comment (ForsakenHarmony) [inline test/e2e/app-dir/app-external/app/optout/action/page.js:14]: same here

**Review by sokra — APPROVED** [review]

File: test/e2e/app-dir/app-external/app/optout/action/actions.js [hunk 3]
```diff
@@ -1,6 +1,7 @@
 'use server'
 
 import { value as dualPkgOptoutValue } from 'dual-pkg-optout'
+import * as mod from 'dual-pkg-optout'
```
Comment (ForsakenHarmony) [inline test/e2e/app-dir/app-external/app/optout/action/actions.js:4]: also probably want to remove this unused import

File: test/e2e/app-dir/app-external/app/optout/action/page.js [hunk 4]
```diff
@@ -10,7 +10,8 @@ export default function Page() {
       <p>{optoutDisplayValue}</p>
       <button
         onClick={async () => {
-          setOptoutDisplayValue(await getDualOptoutValue())
+          const value = await getDualOptoutValue()
```
Comment (ForsakenHarmony) [inline test/e2e/app-dir/app-external/app/optout/action/page.js:13]: Could just revert these two files to what they were before

**Pull request merged**